api/client/src/main/java/javax/websocket/ContainerProvider.java

```diff
@@ -39,20 +51,51 @@
      * @return an implementation provided instance of type WebSocketContainer
      */
     public static WebSocketContainer getWebSocketContainer() {
-         WebSocketContainer wsc = null;
-        for (ContainerProvider impl : ServiceLoader.load(ContainerProvider.class)) {
+        Iterable<ContainerProvider> providers = getProviders();
+        if (!providers.iterator().hasNext()) {
+            throw new RuntimeException("Could not find an implementation class.");
+        }
+        WebSocketContainer wsc = null;
+        for (ContainerProvider impl : providers) {
             wsc = impl.getContainer();
             if (wsc != null) {
                 return wsc;
             } 
         }
-        if (wsc == null) {
-            throw new RuntimeException("Could not find an implementation class.");
-        } else {
-            throw new RuntimeException("Could not find an implementation class with a non-null WebSocketContainer.");
-        }
+        throw new RuntimeException("Could not find an implementation class with a non-null WebSocketContainer.");
     }
  
+    private static Iterable<ContainerProvider> getProviders() {
+        ClassLoader loader = Thread.currentThread().getContextClassLoader();
+        SoftReference<Iterable<ContainerProvider>> providersRef;
+        Iterable<ContainerProvider> providers;
+        // WeakHashMap is not thread-safe. Sync here because a read concurrent with a write can throw an Exception. 
+        synchronized (CACHE) {
+            providersRef = CACHE.get(loader);
+        }
+        if (providersRef != null) {
+            providers = providersRef.get();
+            // Use providers from this point as the SoftReference checked above may be cleared at any point.
+            if (providers != null) {
+                return providers;
+            }
+        }
+        synchronized (CACHE) {
+            providersRef = CACHE.get(loader);
+            if (providersRef == null || (providers = providersRef.get()) == null) {
+                // Use providers from this point as the SoftReference checked above may be cleared at any point.
+                Set<ContainerProvider> set = new HashSet<ContainerProvider>();
+                for (ContainerProvider provider : ServiceLoader.load(ContainerProvider.class, loader)) {
+                    set.add(provider);
```

Comment:
While ServiceLoader loads and caches providers lazily, this is going to instantiate all available providers eagerly - what is the benefit here?

Reply:
We could modify this so it only instantiates the providers until it finds one that returns a WebSocketContainer (like the original code) and then cache that single ContainerProvider. That assumes that once a working ContainerProvider has been found, it will remain working while it remains in the cache. The advantage of instantiating and caching all the providers is that it avoids this assumption. I think the assumption is safe but I'm not certain so I lean towards the current implementation.